pkg/redfish/redfish-foxconn.go

```diff
@@ -0,0 +1,323 @@
+func discoverFoxconnENetInterfaces(s *EpSystem) {
+	//////////////////////////////////////////////////////
+	// Parse /redfish/v1/Systems/system/Oem/Insyde/Ncsi
+
+	path := s.SystemRF.OEM.InsydeNcsi.Ncsi.Oid
+
+	url := s.epRF.FQDN + path
+	jsonData, err := s.epRF.GETRelative(path)
+	if err != nil || jsonData == nil {
+		s.LastStatus = HTTPsGetFailed
+		return
+	}
+	s.LastStatus = HTTPsGetOk
+
+	var n InsydeOemNcsiCollection
+	if err := json.Unmarshal(jsonData, &n); err != nil {
+		errlog.Printf("Failed to decode %s: %s\n", url, err)
+		s.LastStatus = EPResponseFailedDecode
+	}
+
+	if n.MembersOCount < 1 {
+		errlog.Printf("No Ncsi members detected")
+		return
+	}
+
+	//////////////////////////////////////////////////////
+	// Parse each /redfish/v1/Systems/system/Oem/Insyde/Ncsi/# entry
+
+	for _, ncsiMember := range n.Members {
+		path := ncsiMember.Oid
+
+		url := s.epRF.FQDN + path
+		jsonData, err = s.epRF.GETRelative(path)
+		if err != nil || jsonData == nil {
+			s.LastStatus = HTTPsGetFailed
+			return
+		}
+		s.LastStatus = HTTPsGetOk
+
+		var nm InsydeOemNcsiMember
+		if err := json.Unmarshal(jsonData, &nm); err != nil {
+			errlog.Printf("Failed to decode %s: %s\n", url, err)
+			s.LastStatus = EPResponseFailedDecode
+		}
+
+		//////////////////////////////////////////////////////
+		// Parse each /redfish/v1/Systems/system/Oem/Insyde/Ncsi/# package member element
+		// Have only ever seen one package but we should iterate anyway in case that ever changes
+
+		for _, nmPkg := range nm.Package {
+
+			//////////////////////////////////////////////////////
+			// Parse /redfish/v1/Systems/system/Oem/Insyde/Ncsi/#/Package/#
+
+			path = nmPkg.Oid
+
+			url = s.epRF.FQDN + path
+			jsonData, err = s.epRF.GETRelative(path)
+			if err != nil || jsonData == nil {
+				s.LastStatus = HTTPsGetFailed
+				return
+			}
+			s.LastStatus = HTTPsGetOk
+
+			var p InsydeOemPackage
+			if err := json.Unmarshal(jsonData, &p); err != nil {
+				errlog.Printf("Failed to decode %s: %s\n", url, err)
+				s.LastStatus = EPResponseFailedDecode
+			}
+
+			//////////////////////////////////////////////////////
+			// Parse each /redfish/v1/Systems/system/Oem/Insyde/Ncsi/#/Package/#.PackageInfo[]
+
+			for j, pi := range p.PackageInfo {
+				// Only process if there is a MAC address
+				if pi.MACAddress != "" {
+					s.ENetInterfaces.Num++
+
+					ei := NewEpEthInterface(s.epRF, ncsiMember.Oid, s.RedfishSubtype, nmPkg, s.ENetInterfaces.Num)
+
+					ei.EtherIfaceRF.Oid = path
+					ei.EtherIfaceRF.MACAddress = pi.MACAddress
+					ei.EtherIfaceRF.Description = FOXCONN_ETH_INT_DESCRIPTION
+
+					// ID = "ncsi-" + ncsi number + "-" + package number + "-" + channel index
+					ei.EtherIfaceRF.Id = "ncsi" + nm.Id + "-p" + p.Id + "-c" + fmt.Sprint(j)
+
+					// According to Foxconn, this is the only unique identifier for the primary ethernet
+					if nm.VersionId.PCIDID == FOXCONN_PRIMARY_ETH_INT_PCIDID {
```

Comment:
This is the method to uniquely identify the primary ethernet interface given to us by Foxconn in PRDIS-38.  Their last comment did not leave me with the greatest confidence that it will be unique for the lifetime of the Paradise product.  My worry is if they use a new part in the future that may get a new PCIDID.  I did leave a breadcrumb behind in smd's log below if we ever detect a different PCIDID being used.  That too has uniqueness issues though I'll comment on below.